content/blog/2022-06-01-Jupyter-in-cloud-with-TPI.md

```diff
@@ -0,0 +1,379 @@
+
+That's where ngrok comes in, which will generate URLs we can navigate to in
+order to access our environment. Run `terraform refresh` to get these URLs once
+TPI has finished provisioning the instance:
+
+```yaml
+Outputs:
+
+urls = [
+  "Jupyter Lab: https://8c62-54-173-120-3.ngrok.io/lab?token=...",
+  "Jupyter Notebook: https://8c62-54-173-120-3.ngrok.io/tree?token=...",
+  "TensorBoard: https://6d52-54-173-120-3.ngrok.io",
+]
+```
+
+<admon type="info">
+
+It may take a little while for TPI to provision the instance. If
+`terraform refresh` returns `urls = []` as its output, just wait a few minutes
+and rerun it.
+
+</admon>
+
+Once we have accomplished everything we want to do in our notebook, we can
+terminate the instance with `terraform destroy`. This ensures that we won't have
+idle instances racking up our credit card bill. Terraform will download the
+files stored there (such as notebooks) to the shared directory on our local
+machine when terminating the instance.
+
+# Under the hood
+
+"Alright, easy enough, but what is actually happening?"
+
+The major point of interest in the repository is
+[`main.tf`](https://github.com/iterative/blog-tpi-jupyter/blob/aws/main.tf).
+This file contains the specifications Terraform needs to provision our instance.
+Let's take a look at what each part of this configuration does:
+
+```
+terraform {
+  required_providers { iterative = { source = "iterative/iterative" } }
+}
+provider "iterative" {}
+```
+
+This first part of the configuration simply tells Terraform that we will be
+using TPI. We can consider it analogous to a package import in Python.
+
+```
+resource "iterative_task" "jupyter_server" {
+  spot      = 0             # auto-priced low-cost spot instance
+  timeout   = 24*60*60      # force shutdown after 24h
+  disk_size = 125           # GB
+  machine   = "m+t4"        # m/l/xl (CPU), +k80/t4/v100 (GPU)
+  image     = "nvidia"      # or "ubuntu"
+
+  # cloud-specific config
+  cloud     = "aws"         # see `git checkout generic` branch for: gcp, az, k8s
+
+  # blank means extract from local env vars
+  environment = { NGROK_TOKEN = "", TF_CPP_MIN_LOG_LEVEL = "1", QUIET = "1", GITHUB_USER = "username" }
+  storage {
+    workdir = "shared"
+    output  = "."
+  }
+```
+
+The second part of the configuration is perhaps the most interesting. Here we
+provide the specifications for the instance Terraform should provision for us.
+In this case, we would like a medium CPU and an NVIDIA T4 GPU, along with a
+125GB disk. Because machine types vary between cloud vendors,
+[TPI does some translation](https://registry.terraform.io/providers/iterative/iterative/latest/docs/resources/task#machine-type)
+from generic types (e.g., `s`/`m`/`l`/`xl`) to specific cloud machine types.
+This allows us to generalize these configurations and quickly switch from AWS to
+Azure, for example.
+
+Of particular interest here is the `spot = 0`, which tells TPI to provision spot
```

Comment:
given the blog's tone & level of detail in explanations, it implies a beginner target audience. Should thus explain briefly what spot/preemptible means. Pros/cons, give references.

It's also the No. 1 USP of TPI (auto-recovery from spot instance interrupts).